### ISSUE TITLE
DRA: structured parameters: network-attached resource

### ISSUE THREAD

**Issue by pohly** [kubernetes/kubernetes#124042]
### What would you like to be added?

A network- or fabric-attached resource is something that can be attached to a node when needed, but which isn't visible to kubelet before it is attached.

For structured parameters, this leads to some additional challenges:
- A ResourceSlice must be able to describe which nodes the resource instances can be attached to. More than one slice might be needed, so the "named resources" model must support more than one slice per driver. Helper code for publishing and syncing ResourceSlice objects in a driver would be useful.
- Pod scheduling cannot get delayed (doing so could lead to partial allocation where native resources are gone by the time that the resource is attached), so a network-attached resource needs to be allocated and only then does the kubelet plugin of the driver or some other, driver-specific control plane component "attach" the instance. Until then, the pod needs to wait in a retry loop in kubelet. Not all drivers will need to do some real attach operation.
- There may be attach limits per node (PCI switch for composable dissagregated infrastructure) and/or maximum number of overall users per resource instance (Akri).

/cc @diconico07
/sig node
/triage accepted
/priority important-soon

### Why is this needed?

Akri needs this for resources like IP cameras: https://docs.google.com/document/d/1XNkTobkyz-MyXhidhTp5RfbMsM-uRCWDoflUMqNcYTk/edit#heading=h.wjb5cok1vm2x

Composable disaggregated infrastructured needs it for GPUs connected to a node on demand:
https://kccnceu2024.sched.com/event/1ZPDw/iown-bof-challenges-of-kubernetes-for-composable-disaggregated-computing-naoki-oguchi-fujitsu-hidetsugu-sugiyama-red-hat-clara-li-intel-ryosuke-kurebayashi-ntt

**Comment by MikeZappa87** [comment]
@aojea

**Comment by pohly** [comment]
/milestone 1.31

I intend to cover the first point (= publishing ResourceSlice) in 1.31. Considering attach limits may have to come later.

**Comment by hase1128** [comment]
@pohly 
I considered the direction of collaboration between CDI and DRA (+ structured parameters).
Please let me know your opinion.

Initially (when I talked to you at KubeCon Paris), I considered calling standardized CDI functions from the DRA controller so that DRA could also retrieve resources on the fabric.
However, it was found that the collaboration between DRA and CDI has the following issues.

Issue 1 (structured parameters):
  - In CDI, it is difficult to publish the resources available to nodes in advance.
  - Information on the number of resources and model names can be published, but it is difficult to specify in advance which resources will actually be selected.
  - Therefore, some restrictions are needed on CDI usage/operation.

Issue 2 (Cluster Autoscaler):
  - Cluster Autoscaler (CA) will no longer be able to operate properly
  - CA scales in/out nodes with the same spec on a node-by-node basis
  - On the other hand, CDI scales down/up by device in nodes
  - As a result, it becomes difficult to keep the same spec in node groups, making it impossible to make correct scale decisions.

Issue 3 (Triggering configuration changes):
  - In DRA, resources are increased/decreased when Pods are created/deleted.
  - Automatically increase or decrease Pods (Resource Claims) according to load by using Horizontal Pod Autoscaler
  - This means that by working with CDI, we can automatically attach/detach resources on the fabric depending on the load.
  - However, with DRA, it is difficult to increase/increase nodes and resources in advance in a planned manner.

Considering the above issues and the roles of CDI and DRA, I believe there are two main directions.

Direction 1:
  - Allow DRA to also acquire resources on fabric by calling standardized CDI functions from DRA contorller
  - However, issues 1 and 2 need to be resolved. Issue 3 cannot be addressed
  - This direction also requires modifying each vendor's resource driver.

Direction 2:
  - Use DRA after CDI attaches resources on the fabric to nodes in advance
  - In other words, direction 1 includes CDI operations in DRA processing, but direction 2 means CDI and DRA operate independently.
  - For example, when adding a resource, perform the following two steps.
    - 1. CDI adds resources to the node in advance. Update node resource slice as needed
    - 2. DRA(+structured parameters) uses resources added to the node
  - With this idea, there is no need to address issue 1. 
  - Also, regarding CA in issue 2, we does not use CA, but instead requires users to request CDI to adjust the number of nodes and specifications themselves.
  - It is possible to deal with planned configuration changes like issue 3.
  - Each vendor's resource driver does not need to be modified.

![CDI_and_DRA](https://github.com/kubernetes/kubernetes/assets/12468999/d2597863-55d7-49a0-a4eb-cdf102b487bb)

Direction 2 is divided into two steps as shown in the diagram above.
Step1: "define the shape of nodes"
  - Define the CRD for the cluster configuration (number of nodes, type and number of resources connected to the nodes)
  - Modify the contents of the CR manually or automatically by the user or some monitoring tool
  - CDI attaches/detaches resources from the fabric to nodes according to the contents of CR

Step2: "create a pod with resource requests"
  - This is the same as classic DRA. We can also use structured parameters.

Step1 and Step2 are independent operations, but if we change the node configuration in Step1, we need to update the node resource slice, so we still need to discuss how CDI and DRA work together.

In any case, first of all, please let me know your opinion on this direction.
I thought about discussing it in the Device Management WG, but since the topic is quite limited, I decided to use this issue.
If you have any other good places to discuss please let me know

**Comment by hase1128** [comment]
Thank you for your comment. Sorry for the late reply.
I am very sorry about the word CDI. From now on I use "composable disaggregated infrastructure".

Also, I think my comment above wasn't clear enough and may have confused you. I will state my thoughts again below.

I think there are two directions and I think it is better to realize from direction2
I believe you have commented on only direction1, which is described below(and the above comment).

Direction2
Allow composable disaggregated infrastructure and DRA to operate independently without directly linking them. (I didn't tell you this idea at kubecon.)
Specifically, please refer to Direction2 in the comment above.

Direction1
This is an idea I told you about at KubeCon. This is how to call standardized CDI functions from the Classic DRA controller.
However, there are three issues mentioned above comment.
The following is an issue for Direction1, so it is not relevant for Direction2

>>Issue 1 (structured parameters):
>> - In CDI, it is difficult to publish the resources available to nodes in advance.

>They don't need to be published in advanced. The intend is that `NodePrepareResource" makes an allocated resource available on a node once it is needed.

Regarding "publish", what I want to mean is that user must specify in advance the resources on the fabric that can be connected to a certain node in the resource slice. This must be done in advance, before creating the Pod (resource cliam). But I thought it was a problem because it's difficult to determine in advance which resources on the fabric will be connected to a node. Please let me know if I misunderstand.

>>Issue 3 (Triggering configuration changes):
>
>With "classic" DRA (= KEP https://github.com/kubernetes/kubernetes/issues/3063) that would have been the role of the DRA >driver control plane controller. With structured parameters, there's only the kubelet plugin which could run some custom code.
>
>I know that we talked about how to integrate arbitrary DRA drivers into Composable Disaggregated Infrastructure without having to modify the drivers. I am not sure whether that will be possible. At the very least, drivers probably will have to add some additional support when used in that mode.
>
>How to do cluster reconfiguration through some standardized API is the other half of this problem. I don't know how what that should look like and consider it out of scope for this issue. This needs to be discussed elsewhere, probably in an entirely separate design document.

What I wanted to say here is that the trigger for DRA to acquire resources on the network or fabric is when a Pod (Resource Claim) is created.
If we know in advance that a large amount of resources will be required, we need to connect the resources on the fabric to the nodes in advance and scale them up sufficiently. We just use DRA after that.
When dealing with such cases, I believe that using an Operator for composable disaggregated infrastructure like Direction2 is effective.


I think Operator and standard APIs for composable disaggregated infrastructure may out of scope, but I would like to hear your opinion on these directions. (I am concerned about the need to think about it in two directions like this, and whether the idea of each direction makes sense.)

**Comment by pohly** [comment]
> user must specify in advance the resources on the fabric that can be connected to a certain node in the resource slice

This is not the something that an (end) user should have to do. I see that as the role of the cluster admin, perhaps with help from whatever management tooling they use to bring up hardware.

> If we know in advance that a large amount of resources will be required, we need to connect the resources on the fabric to the nodes in advance and scale them up sufficiently.

That's the kind of resource planning that an admin needs to do based on expected load. Automation like "cluster autoscaling" help here in the cloud, but not when it comes to actually installing hardware. DRA doesn't cover this.

**Comment by hase1128** [comment]
> > user must specify in advance the resources on the fabric that can be connected to a certain node in the resource slice

> This is not the something that an (end) user should have to do. I see that as the role of the cluster admin, perhaps with help from whatever management tooling they use to bring up hardware.

I assumed that user was also a cluster administrator. Anyway, I believe it is difficult to specify in advance the resources on the fabric that can be connected to a certain node into the resource slice, because it is difficult to know in advance which resources in the fabric will be allocated to which nodes.(We may need some kind of management tool like you say)

By the way, let me confirm the following three things.
1:
Looking at the source below, it seems that the current resource slice is automatically updated by the DRA plugin in kubelet. (DRA plugin is monitoring the resource status of the node and updating the resource slice by notifying kubelet of any changes)
https://github.com/kubernetes/kubernetes/blob/b722d017a34b300a2284b890448e5a605f21d01e/pkg/kubelet/cm/dra/plugin/noderesources.go#L377
https://github.com/kubernetes/kubernetes/blob/b722d017a34b300a2284b890448e5a605f21d01e/pkg/kubelet/cm/dra/plugin/noderesources.go#L205

In other words, if we attach a resource on the fabric to a node, is the resource slice automatically updated?

2:
I suppose some management tool could update the resource slice, but if the actual device state of the node doesn't change, DRA plugin in kubelet might restore the resource slice state.
Is this a correct understanding?

3:
The resource slice looks like an "**actual state**" indicating the device state of the current node.

We want to define a custom resource for the "**desired state**" of the device state of nodes in terms of a Composable Disaggregated Infrastructure.

As a whole, I imagine the following processing flow:. (This is the architecture of Direction2 in the link below.).
https://github.com/kubernetes/kubernetes/issues/124042#issuecomment-2065673058)

1.   Step1: Cluster administrator updates "**desired state**" (e.g., increases the number of GPUs connected to a node)
2.   Step2: K8s Operator for Composable Disaggregated Infrastructure monitors "**desired state**". If there is a difference between "**desired state**" and "**actual state(resource slice)**", actually get/return the device from the fabric
3.   Step3: kubelet's DRA plugin automatically updates the resource slice (**actual state**).
4.   Step4: (end) user uses DRA(includes structure parameters).

I think Operator for Composable Disaggregated Infrastructure is out of scope for DRA (Be developed by each Composable Disaggregated Infrastructure provider)

However, does it make sense to define a custom resource for the **desired state** of device state in nodes as a DRA? I would like your opinion.

**Comment by pohly** [comment]
> Looking at the source below, it seems that the current resource slice is automatically updated by the DRA plugin in kubelet. 

That will change for network-attached resources. ResourceSlices for those will neither be published nor updated by kubelet.

> I suppose some management tool could update the resource slice

For the scope of this particular issue, the DRA driver is expected to have a control-plane component which does that.

> The resource slice looks like an "actual state" indicating the device state of the current node.

It's "available node-local resources". "State" would also include things like "health" or "in-use". We have thought about adding a status for that, but haven't done it yet.

**Comment by hase1128** [comment]
> That will change for network-attached resources. ResourceSlices for those will neither be published nor updated by kubelet.

Does network-attached resource mean resources that are not connected to a node?
In the case of fabric, kubelet can not update resources that are not yet connected to a node. However, once connected to a node, it will be updated by the kubelet. Is this correct?

> For the scope of this particular issue, the DRA driver is expected to have a control-plane component which does that.

So if the state of a resource slice changes without any actual configuration changes, the kubelet may revert the resource slice back to its original state? 

> It's "available node-local resources". "State" would also include things like "health" or "in-use". We have thought about adding a status for that, but haven't done it yet.

I see. ”health” and "in-use" state looks useful. And I would like to suggest adding desired state in terms of Composable Disaggregated Infrastrucgture. But if resource slice means "available node-local resources", is it unnatural to add desired state to resource slice?(this desired state includes resources that are not available yet (such as resources that the Operator for Composable Disaggregated Infrastructure will connect to nodes from now on).

**Comment by pohly** [comment]
> Does network-attached resource mean resources that are not connected to a node?

Yes, but that resource then is always treated like that, even when it is temporarily attached to a node. A better way to think of this issue really is "IP camera", not "GPU via PCI switch".

> In the case of fabric, kubelet can not update resources that are not yet connected to a node. However, once connected to a node, it will be updated by the kubelet. Is this correct?

No. Kubelet and the DRA driver will have to be aware that the device is tracked as "network-attached" and not interfere with the controller which publishes the ResourceSlice for it.

> But if resource slice means "available node-local resources"

It does in the current KEP merely because that is what can be expressed. This issue is proposing to change that so that a ResourceSlice can also describe a non-node-local resource.

**Comment by SubhasmitaSw** [comment]
Hello @hase1128 @pohly 👋🏻  Just a gentle reminder the `code freeze starts at 02:00 UTC on Wednesday 10th July 2024 (about 4 weeks from now)`, and while there is still plenty of time, we want to ensure that each issue has a chance to be included on time.

As the issue is tagged for 1.31, is it still planned for this release?

**Comment by sanchita-07** [comment]
Hi @pohly @hase1128 👋, following up to check the status of this issue.

**Reminder**: Code freeze starts `02:00 UTC Wednesday 24th July 2024 / 19:00 PDT Tuesday 23rd July 2024` (about 1.5 weeks from now). Please ensure that the related PRs have both the `lgtm` and `approved` labels before the code freeze.

**Comment by pohly** [comment]
1.31 includes support for network attached devices:
https://github.com/kubernetes/kubernetes/blob/11ace3ad97e8c9506abb5267cdcad2b169b5a870/staging/src/k8s.io/api/resource/v1alpha3/types.go#L119-L136

It does not support additional features like restricting the number of devices accessed by a node. I'm leaning towards closing this issue and opening another that is more focused on future enhancements. On the other hand, this issue already contains additional discussion.

@hase1128: what do you think?

**Comment by hase1128** [comment]
If network attached device is supported, I think this issue can be closed. As you say, this issue contains multiple discussions, so I think it's a good idea to create another issue. I was focusing on using DRA in a fabric environment such as CXL (which I call composable disaggregated infrastructure).

**Comment by Vyom-Yadav** [comment]
/milestone clear
We are way past the code freeze. Please add milestone for the next release cycle if that's the target.

**Comment by pohly** [comment]
The issue had "attach limits" as third point, so let's keep it open instead of creating a new issue. I updated the description to make it clear that it was partially solved in 1.31 already.

The WG Device Management Slack channel or meeting are probably good places to discuss what is missing and how to address those remaining gaps. But further comments here are also possible.

**Comment by hase1128** [comment]
@pohly 
> The issue had "attach limits" as third point, so let's keep it open instead of creating a new issue. I updated the description to make it clear that it was partially solved in 1.31 already.

I have a question.
You said "partially solved in 1.31".

I want to know the detail of that. Is there any option or feature is merged in DRA(k8s1.31)?
If yes, please tell me the URL or source code or document of that.

**Comment by pohly** [comment]
In 1.31, structured parameters got extended so that ResourceSlice can describe non-local resources:
https://github.com/kubernetes/kubernetes/blob/09e5e6269aa410609f996ce6d04ca0ee15420d08/staging/src/k8s.io/api/resource/v1alpha3/types.go#L120-L137

A DRA driver managing non-local devices can advertise them in a ResourceSlice with those fields set instead of `nodeName` and the scheduler will correctly allocate and schedule pods such that the final node has access to the devices.

**Comment by hase1128** [comment]
When using a non-node-local device, we can use NodeSelector or AllNodes.
This allows k8s-scheduler to decide which node to deploy the pod.
At this point, I think a process is required to connect the non-node-local device to the node in Composable Disaggregated Infrastructure.

Who will perform this connection process?
For example, will a DRA driver for Composable Disaggregated Infrastructure be required? 
(In that case, I'm concerned about the relationship with other drivers (e.g. NVIDIA's DRA driver).)

Also, even if k8s-scheduler decides where to deploy the pod, the pod will not be Running until the device is actually connected to the node.
Is the process flow below correct?

1.User creates a pod using a ResourceClaim
2.k8s-scheduler decides which node to deploy the pod
3.kubelet starts the pod but it does not become Running
  (Waits until the DRA driver? connects the device to the node)
4. The pod becomes Running

**Comment by hase1128** [comment]
@pohly 
I think you had a conversation at KubeCon NA with my boss(Mr. Oguchi) about dynamic device scaling using DRA in Composable Disaggregated Infrastructure. In a Composable system, ResourceSlice must be updated dynamically as devices are attached/detached to nodes. We are considering doing this with Composable DRA kubelet plugin.

I also heard that there is a method that Composable DRA kubelet plugin sends a request to NVIDIA's DRA kubelet plugin by unix domain socket. We can update ResourceSlice If NVIDIA DRA kubelet plugin has a handler of such request.
Including this idea, I am considering a method of realization. I have a few things I would like to consult with you.

1. Is it difficult for the DRA kubelet plugin itself to periodically get the latest device list and update ResourceSlice?
   I assume that each vendor's DRA kubelet plugin will periodically run the following PublishResources:
   https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/dynamic-resource-allocation/kubeletplugin/draplugin.go#L62

   For example, if we expect to exclude certain GPUs from ResourceSlice, such as when a GPU fails, It is better to update ResourceSlice periodically.

2. Do I always need to use composable DRA kubelet plugin to update ResourceSlice?
   In addition to the above mentioned method in which the vendor(NVIDIA) DRA periodically updates the ResourceSlice on its own, is there any other way?
   For example, If we can generate Resources to pass to PublishResources(ctx context.Context, resources Resources), is it possible to publish and update ResourceSlice by other than composable or vendor's DRA kubelet plugin?

**Comment by pohly** [comment]
The idea that I outlined at KubeCon was roughly like this:
- A central controller provided by the composable infrastructure is responsible for publishing ResourceSlices with GPUs connected via a switch, *not* a kubelet plugin. These ResourceSlices must use a node selector which describes where those devices can be attached.
- On the node, the composable infrastructure provides a DRA driver which acts as proxy between the kubelet and vendor DRA driver(s). In NodePrepareResources, that proxy must first attach the device, then call the vendor driver. NodeUnprepareResources detaches again.

In this setup, the vendor driver must not publish ResourceSlices, otherwise the same device would appear in two different ResourceSlices. It must be able to handle hotplugging.

To implement such a DRA driver proxy, the proxy must implement the [kubelet plugin registration API](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/pluginregistration/v1). To the vendor DRA driver it then looks like it registers with the kubelet. The deployment needs to be a bit different (different paths for Unix Domain sockets).

**Comment by hase1128** [comment]
Thank you for your idea. I have some questions.

> A central controller provided by the composable infrastructure is responsible for publishing ResourceSlices with GPUs connected via a switch, not a kubelet plugin. These ResourceSlices must use a node selector which describes where those devices can be attached.

Q1:
In this case, is a central controller not included DRA driver?(because this is not kubelet plugin)
In my understanding, DRA driver is basically composed of kubelet plugin. Other than kubelet plugin, there is only an admission webhook.
https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4381-dra-structured-parameters#components

Q2:
If a user specifies a vendor driver such as nvidia in the ResourceClaim, K8s scheduler will not refer to the above ResourceSlice exposed by a central controller. Is this correct understanding?
(I thought K8s scheduler will check for a ResourceSliceSpec that specifies the same DRA driver as the one specified via the ResourceClaim.)

Q3:
Can a central controller expose a ResourceSlice using AllNodes?
In this case, it means that the GPU can be connected to any node. Therefore, there will be competition between nodes to acquire the GPU, but I am considering using AllNodes so that the node to which the GPU is connected can be changed depending on the node's usage, rather than limiting the node to which the GPU is connected in advance.

**Comment by pohly** [comment]
> Q1: In this case, is a central controller not included DRA driver?

A DRA driver for network-attached devices is expected to have a control-plane component in addition to the kubelet plugin. In your case, you target "wrapping node-local DRA drivers" - those won't know how to discover and publish devices that only exist in the cluster.

> Q2: If a user specifies a vendor driver such as nvidia in the ResourceClaim, K8s scheduler will not refer to the above ResourceSlice exposed by a central controller. Is this correct understanding?

No, this should work fine. Suppose the driver normally publishes node pools with driver name "dra.example.com". The cluster pool then can use that same name and a DeviceClass which checks for that name will match, so the scheduler would transparently pick either a local GPU or a network-attached one, depending on what is available.

> Q3: If we restart vendor DRA driver for some reasons after GPU is attached via PCIe switch. Vendor DRA kubelet plugin create ResourceSlice which contains the above GPU.

This is where you need the ability to influence how the DRA driver publishes devices. You probably want a `--disable-publish-resource-slices` flag and use that on nodes where the driver is meant to be used only with dynamically attached devices. Mixing local devices with dynamically attached ones would be harder, but perhaps not necessary?

**Comment by hase1128** [comment]
I understand.
I think such flag(--disable-publish-resource-slices) is required for us. I'll consider into the overall architecture a bit more. 
In a composable server, there are few cases where devices such as GPUs are directly installed on the server.

For now, I believe if we implement this flag in k8s kubelet plugin, the vendor's DRA kubelet plugin won't need to be modified.

**Comment by KobayashiD27** [comment]
@pohly 

When creating a composable DRA driver, we are considering providing a DeviceClass like the one below to enable transparent selection between node-local NVIDIA resourceSlices and composable resourceSlices on the fabric. Does this align with your perspective?

```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: DeviceClass
name: composable.example.com
spec:
  selectors:
  - cel:
      expression: device.driver == "composable-driver.example.com" || device.driver == "gpu.nvidia.com"
```

We aim to allow users to specify their requirements without needing to be aware of whether the resource is node-local or on the fabric. Therefore, we believe it is preferable to specify `DeviceClassName="gpu.nvidia.com"` rather than `DeviceClassName="composable.example.com"`. 

In your opinion, what kind of DeviceClass and resourceSlice design would be suitable to create a system where users request NVIDIA devices, and the scheduler transparently selects from both node-local and fabric-based devices? While the specific DeviceClass implementation is ultimately determined by the cluster administrator, we would appreciate your insights on what configuration would best suit a DRA.

**Comment by pohly** [comment]
The way I had envisioned it, you wouldn't need such a special class. When your composable controller publishes ResourceSlices, it does so with `driver: gpu.nvidia.com`. Then the same DeviceClass matches attached and local devices.

**Comment by KobayashiD27** [comment]
@pohly 

My understanding is that when the composable controller exposes a resourceClass with `driver: gpu.nvidia.com`, the kubelet DRA driver called when that resourceSlice is selected is `gpu.nvidia.com`.

If the device exposed by the resourceSlice is a fabric-attached device, it needs to be attached to the node. However, the DRA kubelet plugin provided by NVIDIA does not include the process for attaching fabric-attached devices to the node.

Therefore, I am considering the following process flow. Does this align with your expectations?

1. NodePrepareResources() of the DRAPlugin is called from kubelet.
2. The NodePrepareResources() of the `composable-driver.example.com` kubelet DRA plugin executes the process to attach the device to the node.
3. The `composable-driver.example.com` kubelet DRA plugin calls the NodePrepareResources() of the `gpu.nvidia.com` kubelet DRA plugin.

I believe a special DeviceClass, as mentioned in the example, is necessary for this process.

I understand that when the composable controller exposes a ResourceSlice using `driver: gpu.nvidia.com`, the same DeviceClass can transparently select between fabric-attached devices and local devices. However, in that case, I believe the device processing will be handled by the `gpu.nvidia.com` kubelet DRA plugin.

If we want to execute composable-specific processing for device attachment, where do you think this processing can be called from? Previously, you suggested a proxy. In my view, the proxy would involve the composable driver being called from the kubelet's NodePrepareResources() call, and then the composable driver calling the NVIDIA driver.

**Comment by pohly** [comment]
> My understanding is that when the composable controller exposes a resourceClass with driver: gpu.nvidia.com, the kubelet DRA driver called when that resourceSlice is selected is gpu.nvidia.com.

The idea was that only the proxy registers with the kubelet, not the original vendor driver. Then the kubelet has only one driver with the name "gpu.nvidia.com", and that is the proxy.

**Comment by KobayashiD27** [comment]
@pohly 
Thank you for explaining your proxy idea. I understand it better now.

Using this proxy idea, how can we utilize the functionalities of the original vendor driver?

Specifically, I would like the composable driver to handle the process of attaching fabric-attached devices to the node, while using the original vendor driver's processes for device initialization and CDI preparation.

Ideally, the original vendor driver would also be running, but it would not accept requests directly from the kubelet. Instead, all requests would go through the composable driver acting as a proxy. However, the original vendor driver performs registration with the kubelet as part of its initialization process, so it cannot be used as-is.

Could you provide any ideas or reference implementations for achieving this approach?

**Comment by pohly** [comment]
> However, the original vendor driver performs registration with the kubelet as part of its initialization process, so it cannot be used as-is.

Yes. Some tweaks to the original vendor deployment will be needed. The vendor has to support:
- A configuration option for "don't publish ResourceSlices".
- A configuration for the kubelet registration path: usually it is `/var/lib/kubelet/plugins_registry`, but even that may depend on the Kubernetes installation. When using the proxy, it could be something like `/var/lib/composable-proxy/plugins_registry`.

**Comment by hase1128** [comment]
@pohly 
> Yes. Some tweaks to the original vendor deployment will be needed. The vendor has to support:

This is one of the issue needs to be solved for us.
As the second issue, I would like to discuss again how to manage resourceslice.

> This is where you need the ability to influence how the DRA driver publishes devices. You probably want a --disable-publish-resource-slices flag and use that on nodes where the driver is meant to be used only with dynamically attached devices. Mixing local devices with dynamically attached ones would be harder, but perhaps not necessary?

Recently, I am considering how to manage both of node-local resources and non-node-local resources.
In composable architecture, there are few cases where devices such as GPUs are directly installed on the server.
However, we may need a node-local ResourceSlice in the following cases:
 - When a composable server is created, several GPUs are attached to it from the beginning (separate from the GPUs that are attached/detached when a Pod is created or deleted by DRA).
 - When using non-composable servers at the same time

Therefore, I propose to manage resource slices as follows.
 - Node-local resource slices: Created and managed by the vendor's DRA driver
 - Non-node-local resource slices on the fabric: Created and managed by the composable DRA driver

If a GPU is attached to a node by creating a pod, the GPU moves from the non-node-local resource slice to the node-local resource slice.
And if a GPU is detached from a node by deleting a pod, the GPU moves from the node-local resource slice to the non-node-local resource slice.

I have described a simple process flow in P1 to P4 below.
Is this idea reasonable?
https://docs.google.com/presentation/d/1S4t1U2FnUoeY9G68WUgg2L30AyPMMfe-2i9slSiDO0s/edit?usp=sharing

**Comment by pohly** [comment]
> Therefore, I propose to manage resource slices as follows.
> 
> *  Node-local resource slices: Created and managed by the vendor's DRA driver
> *  Non-node-local resource slices on the fabric: Created and managed by the composable DRA driver

This is compatible with the DRA design. Compared to what I proposed, some additional support from a vendor will be needed. Instead of "publish ResourceSlice on/off", there has to be a configuration option for "publish ResourceSlice for some devices" with a description of what those are.

> If a GPU is attached to a node by creating a pod, the GPU moves from the non-node-local resource slice to the node-local resource slice.

This is a very intrusive change to how DRA works. It breaks several existing APIs, including the ResourceClaim allocation status. Let's avoid this, please.

**Comment by KobayashiD27** [comment]
> > However, the original vendor driver performs registration with the kubelet as part of its initialization process, so it cannot be used as-is.
> 
> Yes. Some tweaks to the original vendor deployment will be needed. The vendor has to support:
> 
> * A configuration option for "don't publish ResourceSlices".
> * A configuration for the kubelet registration path: usually it is `/var/lib/kubelet/plugins_registry`, but even that may depend on the Kubernetes installation. When using the proxy, it could be something like `/var/lib/composable-proxy/plugins_registry`.

Even if two configurations are introduced to the vendor driver, using dra.kubeletplugin.Start() would still result in registration with the kubelet.  (https://github.com/kubernetes/dynamic-resource-allocation/blob/master/kubeletplugin/draplugin.go#L280)

If this understanding is correct, how can we utilize the vendor driver's functionalities from the composable driver? 
Do we need to implement a registration function equivalent to the DRA kubelet plugin's registration in the composable driver?
I wish there is a more affordable way...

**Comment by pohly** [comment]
> using dra.kubeletplugin.Start() would still result in registration with the kubelet. (https://github.com/kubernetes/dynamic-resource-allocation/blob/master/kubeletplugin/draplugin.go#L280)

That depends on the configuration. If the DRA driver is asked to create it's socket somewhere where the kubelet doesn't see then, there's no registration with the kubelet.

> Do we need to implement a registration function equivalent to the DRA kubelet plugin's registration in the composable driver?

Yes.

**Comment by pohly** [comment]
Here's an alternative design.  The `DRAResourceClaimDeviceStatus` feature gate must be set for this, but no further code changes are needed in Kubernetes.

As before, a composite controller publishes ResourceSlices with all devices that can be attached to nodes. All of those devices have `dra.k8s.io/attachable` (name TBD) with `true` as part of their attributes. When such a device gets allocated in a ResourceClaim, that controller is responsible for attaching it to the node. Once that is complete, it has to add [a device status](https://github.com/kubernetes/kubernetes/blob/810e9e212ec5372d16b655f57b9231d8654a2179/staging/src/k8s.io/api/resource/v1beta1/types.go#L1005-L1053) to `claim.status.devices` with a condition `dra.k8s.io/attached` set to `True`. `Data` can be populated with arbitrary information that'll help the DRA driver on the node to identify the device.

On the node, DRA drivers get registered as usual (no proxy!). When asked to prepare a device that they haven't published themselves (= not local), they check the condition in the claim status. If the device is attached, they use the information in the `Data` field to find the right device. If it is not attached yet, the NodePrepareResources call must fail and will be retried. This already works, albeit isn't perfect because the retries add latency.

Tearing down is similar. The composite controller must detect when a ResourceClaim got deallocated and potentially deleted while the device was still attached, then detach the device from the node or lazily do that at a later time. Perhaps the same device will be used immediately on the same node via some other ResourceClaim.

Some cooperation is needed by DRA driver vendors because the driver , but that was inevitable.

**Comment by KobayashiD27** [comment]
@pohly 

Thank you for the suggestion. 
The alternative approach seems to simplify the structure. While it may appear that the `NodePrepareResources()` call fails multiple times until the device is attached, causing the Pod to fail to start several times, it feels more reasonable than asking the vendor to handle complex registration processes.

As we proceed with the design, I would like to hear your opinion on how to handle fabric-attached devices as resourceSlices. Currently, the association between resourceSlices and devices depends on the vendor's implementation. Therefore, the current vendor driver's implementation creates a database of devices managed by the driver at startup and prepares the devices based on that data. From what I can see in the source code, this database does not get updated.

In other words, fabric-attached devices are not recognized at the time of driver startup and thus cannot be prepared. As it stands, I don't think fabric-attached devices can be controlled by the DRA. Is there anything that the DRA can do to address this issue? For example, could we introduce an interface to rescan the node's device information after attaching the device? 

Even with a rescan, the issue of associating resourceSlices with devices remains, but I would like to hear your thoughts.

**Comment by pohly** [comment]
> In other words, fabric-attached devices are not recognized at the time of driver startup and thus cannot be prepared. As it stands, I don't think fabric-attached devices can be controlled by the DRA. 

This is the "Some cooperation is needed by DRA driver vendors because the driver needs to support local and non-local devices (this part was missing in https://github.com/kubernetes/kubernetes/issues/124042#issuecomment-2514545235), but that was inevitable."

> Is there anything that the DRA can do to address this issue?

I can't think of anything. In the DRA design, the drivers own the hardware. I don't think we should reverse that  by making Kubernetes responsible for device discovery and management. It would not work in the general case.

**Comment by shogom2** [comment]
@pohly 

Thank you for your suggestions. The new idea in https://github.com/kubernetes/kubernetes/issues/124042#issuecomment-2514545235 seems more reasonable than implementing a new proxy. I'd like to verify my understanding.

> If it is not attached yet, the NodePrepareResources call must fail and will be retried. This already works, albeit isn't perfect because the retries add latency.

Q1:
I apologize for asking a basic question, but I'd like to clarify my understanding. I believe that even if the NodePrepareResources() call fails, the Pod will remain in a non-error state (e.g., ContainerCreating). This is based on my observation that SyncPod() does not appear to propagate errors from PrepareDynamicResources(). Is this correct?

Q2:
Is the latency you mentioned due to the interval between SyncPod() calls?

If my understanding is incorrect, and the Pod temporarily enters an error state or is rescheduled after a NodePrepareResources() error, then we should consider improvements. Therefore, I would appreciate clarification on these points.

**Comment by pohly** [comment]
Yes to Q1. For Q2 I don't know exactly which interval is used between retrying calls, but `SyncPod` sounds likely.

**Comment by hase1128** [comment]
@pohly 
Let me additionally confirm this [comment](https://github.com/kubernetes/kubernetes/issues/124042#issuecomment-2511913856).

>> If a GPU is attached to a node by creating a pod, the GPU moves from the non-node-local resource slice to the node-local resource slice.

> This is a very intrusive change to how DRA works. It breaks several existing APIs, including the ResourceClaim allocation status. Let's avoid this, please.

I want to define the ResourceSlice managed by the vendor's DRA driver and the Comoposable DRA driver as:

**Vendor DRA driver:**
Manages one type of ResourceSlice:
 1: Node-local resource slices
    These resources are directly installed in a server

**Composable DRA driver:**
Manages two types of ResourceSlices. The following ResourceSlices belong to the same Pool.
 1: Node-local resource slices
    These resources are connected via PCIe/CXL switch
 2: Non-Node-local resource slices
    These resources are present via PCIe/CXL switch and are not yet connected to a server

When managing the above ResourceSlice, is there any problem with moving the device information managed by the Composable DRA driver in conjunction with attaching and detaching devices?
  - On attach: Move device information from Non-Node-Local resource slice to Node-Local resource slice
  - When detaching: reverse of above

As I understand that, AllocationResult of a ResourceClaim needs the following three data of information to identify the device.
  - Pool name
  - Driver name
  - Device name

Therefore, I thought that it is possible to move Device information between Resource Slices managed by the same Pool and driver.
(Vendor DRA drivers must have an option not to expose devices via the Switch)
Is that correct understanding? Or should we avoid moving?

**Comment by pohly** [comment]
> The following ResourceSlices belong to the same Pool: Node-local resource slices [...]  Non-Node-local resource slices

A pool cannot be both local to a node and non-local. The availability of devices in the pool is defined by the pool, not the individual slices.

**Comment by hase1128** [comment]
Understood. Sorry for my lack of understanding of Pool.
By the way, is it possible to move the device not used by Pod (free device not registered in AllocationResult) from Non-Node-Local ResourceSlice to Node-Local ResourceSlice?

This assumes that the cluster administrator has manually controlled the PCIe SW and attached the device to the node.
In this case, I assume that PublishResource function needs to be called again because more devices are available on the node local.

**Comment by pohly** [comment]
> By the way, is it possible to move the device not used by Pod (free device not registered in AllocationResult) from Non-Node-Local ResourceSlice to Node-Local ResourceSlice?

Yes, but you need even more support from the vendor DRA driver because it needs to update its own ResourceSlice(s). It also sounds risky because there is no ordering between those two pools. A race leading to "device listed in both pools" would be possible unless moving a device is carefully orchestrated.

**Comment by hase1128** [comment]
Thank you for your comment. I understand the risks of the move.

I was thinking of moving devices between node-local resources and non-node-local resources (which exist in different pools) managed by the composable DRA driver.

In this case, I need to ask the vendor's DRA driver not to expose resources connected via a PCIe/CXL switch.

**Comment by pohly** [comment]
Moving the device from one pool to another invalidates the claim allocation status: it refers to the device as being part of the non-node-local pool.

**Comment by KobayashiD27** [comment]
@pohly 

To achieve efficient management of fabric devices, we propose adding the following features to the Kubernetes scheduler's DRA plugin. Fabric devices are those that are not directly connected to the server and require attachment to the server for use.
 
### Background and Purpose of the Proposal
 
In the current implementation, fabric devices are attached after the scheduling decision, which leads to the following issues:
- Fabric resources may be contested by other clusters. In scenarios where attachment occurs after scheduling, there is a risk that the resource cannot be attached at the time of attachment, causing the container to remain in the "Container Creating" state.
 
To address this issue, we propose a feature that allows the scheduler to wait for the device to be attached.
 
### Details of the Proposed Features
 
- Add a Ready Flag to ResourceClaim:
   - Add a flag to ResourceClaim that indicates the readiness state of the device. The PreBind phase will be held until this flag is set to "Ready".
 
### Overview of the PreBind() Process
 
- Update ResourceClaim:
   - The scheduler updates the ResourceClaim to notify the vendor's driver that the device needs to be prepared. This process is the same as the existing PreBind.
   - After updating the ResourceClaim, if the flag is set to "Preparing", the completion of the PreBind phase will be held until the flag is set to "Ready".
 
- Monitoring and Preparation by Composable DRA Controllers:
   - Composable DRA Controllers monitor the ResourceClaim. If a ResourceSlice that requires preparation is associated with the ResourceClaim, they perform the necessary preparations. Once the preparation is complete, they set the flag to "Ready".
 
- Completion of the PreBind Phase:
   - Once the flag is set to "Ready", the PreBind phase is completed, and the scheduler proceeds to the next step.
 
### Impact on Existing Operations
 
This modification applies only to fabric devices and does not affect non-fabric devices. The flag will be set to "Preparing" for fabric devices and "Ready" for non-fabric devices, ensuring that existing operations are not impacted.
 
By adding this feature, it becomes possible to wait until fabric devices are ready, ensuring smooth utilization of these devices. This is still a rough idea, and further discussion on the management of ResourceSlice and ResourceClaim is needed in the next steps. However, we would like to first discuss the feasibility of adding such a feature.
 

![image](https://github.com/user-attachments/assets/329d9e27-1151-4793-8da7-15a313ee8e91)

**Comment by pohly** [comment]
> The scheduler updates the ResourceClaim to notify the vendor's driver that the device needs to be prepared. This process is the same as the existing PreBind.

In other words, the claim is marked as being allocated.

> After updating the ResourceClaim, if the flag is set to "Preparing", the completion of the PreBind phase will be held until the flag is set to "Ready".

Implementation detail: how does the scheduler know that it needs to wait? The flag won't be "Preparing" immediately. This probably needs to be a standardized device attribute, like "k8s.io/needs-attach".

> Completion of the PreBind Phase: Once the flag is set to "Ready", the PreBind phase is completed, and the scheduler proceeds to the next step.

So basically you are proposing that the PreBind method should block infinitely until it either gets a "ready" or (presumably) "failed" from the driver?

Why is that state better than completing the pod scheduling and doing the waiting on the node?

- The pod will be considered scheduled and the devices allocated by the scheduler, so resources are blocked with and without this proposal.
- The driver on the node also needs to "wait" for the device because the device will not be ready immediately after the controller has attached it. The local kernel still needs to detect and initialize the new device.
- Waiting in the scheduler would be event-driven, not periodic as in the kubelet. But we may still encounter NodePrepareResource failures depending on the timing because of the previous point.
- The status is a bit more obvious to users and admins because it is part of the API.
- If attaching fails, the scheduler can roll back and schedule the pod elsewhere.

This last point IMHO is the main reason why this could be useful. But without additional information, the scheduler might just pick the same node again where the device cannot be attached. So we first should figure out how the driver can make fabric constraints visible to the scheduler, then consider this proposal as a way to enhance the attach flow.

**Comment by KobayashiD27** [comment]
@pohly 

Thank you for considering my proposal and for your thoughtful response.

> Implementation detail: how does the scheduler know that it needs to wait? The flag won't be "Preparing" immediately. This probably needs to be a standardized device attribute, like "k8s.io/needs-attach".

Yes, I agree that we need to define a standardized flag for fabric devices. My rough idea is to include information in the ResourceSlice indicating that it is a fabric device. The scheduler can then set the flag if it detects that the device is a fabric device.

> So basically you are proposing that the PreBind method should block infinitely until it either gets a "ready" or (presumably) "failed" from the driver?

I believe we can avoid infinite waiting by implementing a timeout mechanism, but essentially, yes.

> Why is that state better than completing the pod scheduling and doing the waiting on the node?

Fabric devices can be used by multiple Kubernetes clusters, not just a single one. Each cluster expose the device as a ResourceSlice, allowing for efficient utilization of the device.

In this scenario, the ResourceSlice representing a same fabric device might be selected in multiple Kubernetes clusters simultaneously. If the attachment fails in one cluster, the pod will remain in a failed state in kubelet.

By having the scheduler wait for the fabric device to be attached, we can reschedule the pod if the attachment fails. This approach is superior because it avoids unnecessary waiting and allows for immediate rescheduling.

Additionally, although I didn't mention it earlier, the controller will stop exposing the device as a ResourceSlice if the attachment fails. This additional information will prevent the scheduler from selecting the same device again.

![image](https://github.com/user-attachments/assets/72526e57-3465-4d65-a7aa-b1c6dd1be439)
![image](https://github.com/user-attachments/assets/700580a8-3d7a-44f3-8dc6-0e5bfe17004b)

**Comment by pohly** [comment]
So we agree that "can reschedule the pod if the attachment fails" is the key argument in favor of doing this.

> Fabric devices can be used by multiple Kubernetes clusters, not just a single one. [ ... ] Additionally, although I didn't mention it earlier, the controller will stop exposing the device as a ResourceSlice if the attachment fails. This additional information will prevent the scheduler from selecting the same device again.

This is also a good point. If we did the attach only on the node, it would be harder to trigger deallocation of the claim and a full retry.

Besides "device gets removed entirely" to prevent picking it again, do we also need additional information like attach limits per node?

This looks ready to be turned into a KEP. Do you want to work on that for Kubernetes 1.33? I can create the issue (then I become the owner or shepherd of it until it reaches GA) or you can do it (but that then is a long-term commitment because transferring ownership of a KEP issue is currently not supported). Either way, it would be your responsibility to write the KEP text and the implementation. I can review.

**Comment by KobayashiD27** [comment]
@pohly 

Thank you. I'm glad you understand the good point of our proposal.

> Besides "device gets removed entirely" to prevent picking it again, do we also need additional information like attach limits per node?

 I believe it would be beneficial to include such details to ensure more efficient resource allocation and avoid potential conflicts.　However, I cannot immediately provide a detailed answer on where and what specific information should be included. Let's discuss the specifics, including the necessity, during the KEP writing process.
In addition, I would like to prioritize node-local devices when selecting devices. For example, rearranging the order of pools when using GetherPool().

>This looks ready to be turned into a KEP. Do you want to work on that for Kubernetes 1.33? 

Yes. I would like to aim for 1.33.

>I can create the issue (then I become the owner or shepherd of it until it reaches GA) or you can do it (but that then is a long-term commitment because transferring ownership of a KEP issue is currently not supported). Either way, it would be your responsibility to write the KEP text and the implementation. I can review.

I have a question regarding the role of the owner. What are the differences between me taking on the owner role versus you taking it on? If there is a significant difference, we may decide that it is best for you to be the owner. Of course, we will handle the writing of the KEP text and the implementation of the features. We would appreciate your cooperation as a shepherd in this process.

Thank you again for your support and collaboration.

**Comment by pohly** [comment]
> In addition, I would like to prioritize node-local devices when selecting devices. For example, rearranging the order of pools when using GetherPool().

That falls under https://github.com/kubernetes/enhancements/issues/4970 which we deferred for now because it is tricky to make generic.

> I have a question regarding the role of the owner. What are the differences between me taking on the owner role versus you taking it on?

The owner has to update the issue description throughout the life of the issue. If you create the issue, work on it for the initial alpha in 1.33, and then no longer have time for it, we won't have such an owner anymore but also cannot transfer the issue. I'll create the issue, it won't make a difference for doing the actual work.

**Comment by KobayashiD27** [comment]
> > In addition, I would like to prioritize node-local devices when selecting devices. For example, rearranging the order of pools when using GetherPool().
> 
> That falls under [kubernetes/enhancements#4970](https://github.com/kubernetes/enhancements/issues/4970) which we deferred for now because it is tricky to make generic.
> 
Thank you very much for providing the helpful KEP. I understand that generalizing the scoring is challenging. However, I am currently considering whether it might be possible to change the order of the pool for device selection so that node-local devices are checked first.
With the scheduler feature changes we are planning, I believe we can generalize the method to distinguish between fabric devices and node-local devices.
Would it be possible to include this feature in the DRA scheduler filter plugin?

> > I have a question regarding the role of the owner. What are the differences between me taking on the owner role versus you taking it on?
> 
> The owner has to update the issue description throughout the life of the issue. If you create the issue, work on it for the initial alpha in 1.33, and then no longer have time for it, we won't have such an owner anymore but also cannot transfer the issue. I'll create the issue, it won't make a difference for doing the actual work.

I have understood about the owner as the issue owner. For management purposes, I have decided it would be best to entrust this to you. Could you please create the issue?

**Comment by pohly** [comment]
https://github.com/kubernetes/enhancements/issues/5007 got created for the "attach devices" aspect.

**Comment by pohly** [comment]
> However, I am currently considering whether it might be possible to change the order of the pool for device selection so that node-local devices are checked first.

Here's an idea how that  could fit into things planned for the near future: as part of https://github.com/kubernetes/kubernetes/issues/128979 I plan to add an "overlay" mechanism where admins can change attributes of devices. This could be extended to influence the priority of devices an/or entire pools. The initial implementation would simply try devices with a higher priority first. Later, this priority could become part of a more complex scoring algorithm.

**Comment by pohly** [comment]
Most of the support for network-attached devices is available (node selector in slices and even per-device). Binding conditions has a stand-alone KEP and the discussion has mostly moved elsewhere. Let's close this issue.

/close
